CODE_STYLE.md

```diff
@@ -0,0 +1,3293 @@
+        # body omitted
+      end
+    end
+  end
+  ```
+
+* <a name="alias-method-lexically"></a>
+  Prefer `alias` when aliasing methods in lexical class scope as the
+  resolution of `self` in this context is also lexical, and it communicates
+  clearly to the user that the indirection of your alias will not be altered
+  at runtime or by any subclass unless made explicit.
+<sup>[[link](#alias-method-lexically)]</sup>
+
+  ```Ruby
+  class Westerner
+    def first_name
+      @names.first
+    end
+
+    alias given_name first_name
+  end
+  ```
+
+  Since `alias`, like `def`, is a keyword, prefer bareword arguments over
+  symbols or strings. In other words, do `alias foo bar`, not
+  `alias :foo :bar`.
+
+  Also be aware of how Ruby handles aliases and inheritance: an alias
+  references the method that was resolved at the time the alias was defined;
+  it is not dispatched dynamically.
+
+  ```Ruby
+  class Fugitive < Westerner
+    def first_name
+      'Nobody'
+    end
+  end
+  ```
+
+  In this example, `Fugitive#given_name` would still call the original
+  `Westerner#first_name` method, not `Fugitive#first_name`. To override the
+  behavior of `Fugitive#given_name` as well, you'd have to redefine it in the
+  derived class.
+
+  ```Ruby
+  class Fugitive < Westerner
+    def first_name
+      'Nobody'
+    end
+
+    alias given_name first_name
+  end
+  ```
+
+* <a name="alias-method"></a>
+  Always use `alias_method` when aliasing methods of modules, classes, or
+  singleton classes at runtime, as the lexical scope of `alias` leads to
+  unpredictability in these cases.
+<sup>[[link](#alias-method)]</sup>
+
+  ```Ruby
+  module Mononymous
+    def self.included(other)
+      other.class_eval { alias_method :full_name, :given_name }
+    end
+  end
+
+  class Sting < Westerner
+    include Mononymous
+  end
+  ```
+
+## Exceptions
+
+* <a name="fail-method"></a>
+  Signal exceptions using the `fail` method. Use `raise` only when catching an
+  exception and re-raising it (because here you're not failing, but explicitly
+  and purposefully raising an exception).
+<sup>[[link](#fail-method)]</sup>
+
+  ```Ruby
+  begin
+    fail 'Oops'
+  rescue => error
+    raise if error.message != 'Oops'
+  end
+  ```
+
+* <a name="no-explicit-runtimeerror"></a>
+  Don't specify `RuntimeError` explicitly in the two argument version of
+  `fail/raise`.
+<sup>[[link](#no-explicit-runtimeerror)]</sup>
+
+  ```Ruby
+  # bad
+  fail RuntimeError, 'message'
+
+  # good - signals a RuntimeError by default
+  fail 'message'
+  ```
+
+* <a name="exception-class-messages"></a>
+  Prefer supplying an exception class and a message as two separate arguments
+  to `fail/raise`, instead of an exception instance.
+<sup>[[link](#exception-class-messages)]</sup>
+
+  ```Ruby
+  # bad
+  fail SomeException.new('message')
+  # Note that there is no way to do `fail SomeException.new('message'), backtrace`.
+
+  # good
+  fail SomeException, 'message'
```

Comment:
While this is common, it still misses structure and good context.

Like we have `logger.info("message", :context => "goes", :here => true)` we should somehow have an exception that includes both a message (what went wrong) and context (state/data during the exception). Not just a message.